src/language/language_type.hbs.rs

```diff
@@ -53,22 +53,44 @@ impl LanguageType {
         ]
     }
 
-    /// Get language from it's file extension.
+    /// Get language from a file path. May open and read the file.
     ///
     /// ```no_run
     /// # use tokei::*;
-    /// let rust = LanguageType::from_extension("./main.rs");
+    /// let rust = LanguageType::from_path("./main.rs");
     ///
     /// assert_eq!(rust, Some(LanguageType::Rust));
     /// ```
-    pub fn from_extension<P: AsRef<Path>>(entry: P) -> Option<Self> {
-        if let Some(extension) = fs::get_extension(entry) {
+    pub fn from_path<P: AsRef<Path>>(entry: P) -> Option<Self> {
+        let filename = fs::get_filename(&entry);
+
+        if let Some(filename) = filename {
+            match &*filename {
+                {{~#each languages}}
+                    {{~#if this.filenames}}
+                        {{~#each this.filenames}}
+                            "{{~this}}" {{~#unless @last}} | {{~/unless}}
+                        {{~/each}}
+                            => return Some({{~@key}}),
+                    {{~/if}}
+                {{~/each}}
+
+                _ => ()
+            }
+        } 
+
+        let extension = fs::get_extension(&entry)
```

Comment:
The canonical CMake filename is `CMakeLists.txt`, for whatever reason. Maybe to make it easier for Windows users to edit?

Similar cases are `Cargo.toml` and `package.json`, although those are more straightforwardly Cargo and JSON files, so I understand why you wouldn't want to count them separately.

Note that the `filename` key matches the entire filename, not just the extension-stripped stem. So in this scheme, `Makefile.bananas` won't be detected as anything, since there's nothing with the `Makefile.bananas` filename, and nothing with the `bananas` extension.